lib/eventsource/index.js

```diff
@@ -0,0 +1,408 @@
+'use strict'
+
+const { pipeline } = require('node:stream')
+const { fetching, finalizeAndReportTiming } = require('../fetch')
+const { HeadersList } = require('../fetch/headers')
+const { Request } = require('../fetch/request')
+const { getGlobalOrigin } = require('../fetch/global')
+const { webidl } = require('../fetch/webidl')
+const { CONNECTING, OPEN, CLOSED, defaultReconnectionTime, ANONYMOUS, USE_CREDENTIALS } = require('./constants')
+const { EventSourceStream } = require('./eventsource-stream')
+const { parseMIMEType } = require('../fetch/dataURL')
+const { MessageEvent, OpenEvent, ErrorEvent } = require('./events')
+const { isNetworkError } = require('../fetch/response')
+const { getGlobalDispatcher } = require('../global')
+
+let experimentalWarned = false
+
+/**
+ * @typedef {object} EventSourceInit
+ * @property {boolean} [withCredentials] indicates whether the request
+ * should include credentials.
+ */
+
+/**
+ * The EventSource interface is used to receive server-sent events. It
+ * connects to a server over HTTP and receives events in text/event-stream
+ * format without closing the connection.
+ * @extends {EventTarget}
+ * @see https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events
+ * @api public
+ */
+class EventSource extends EventTarget {
+  #events = {
+    open: null,
+    error: null,
+    message: null
+  }
+
+  #url = null
+  #withCredentials = false
+  #readyState = CONNECTING
+  #lastEventId = ''
+  #reconnectionTimer = null
+  #request = null
+  #controller = new AbortController()
+  #settings = null
+
+  /**
+   * Creates a new EventSource object.
+   * @param {string} url
+   * @param {EventSourceInit} [eventSourceInitDict]
+   * @see https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface
+   */
+  constructor (url, eventSourceInitDict = {}) {
+    // 1. Let ev be a new EventSource object.
+    super()
+
+    webidl.argumentLengthCheck(arguments, 1, { header: 'EventSource constructor' })
+
+    if (!experimentalWarned) {
+      experimentalWarned = true
+      process.emitWarning('EventSource is experimental, expect them to change at any time.', {
+        code: 'UNDICI-ES'
+      })
+    }
+
+    url = webidl.converters.USVString(url)
+    eventSourceInitDict = webidl.converters.EventSourceInitDict(eventSourceInitDict)
+
+    // 2. Let settings be ev's relevant settings object.
```

Comment:
actually i dont know what the spec means with relevant settings object. Is this right?

Reply:
yes it is, it's basically just default options but worded in a very confusing way